Library/MapleBacon/MapleBacon/Storage/Storage.swift

```diff
@@ -4,70 +4,146 @@
+     
+     - parameter key: the key
+     
+     - returns: the image
+     */
     func image(forKey key: String) -> UIImage?
-    func removeImage(forKey key: String)
-    func clearStorage()
-
+    
+    /**
+     Removes an image with the given key
+     
+     - parameter key: the key
+     */
+    func remove(imageForKey key: String)
+    
+    /**
+     Clears the entire storage
+     */
+    func clear()
 }
 
 public protocol CombinedStorage {
-
     func clearMemoryStorage()
-
 }
 
-public final class MapleBaconStorage {
-
-    private let inMemoryStorage: Storage
-    private let diskStorage: Storage
+public class MapleBaconStorage {
 
+    /// Singleton instance
     public static let sharedStorage = MapleBaconStorage()
+    
+    /// Available storages
+    private static let inMemoryStorage = "mem"
+    private static let diskStorage     = "disk"
+    
+    /// Holds all available storage adapters
+    private var adapters: [String: Storage]
 
+    /**
+     Private init
+     Creates all storage adapters and inits the general storage
+     */
     private init() {
-        inMemoryStorage = InMemoryStorage.sharedStorage
-        diskStorage = DiskStorage.sharedStorage
+        self.adapters = [
+            MapleBaconStorage.inMemoryStorage: InMemoryStorage.sharedStorage,
+            MapleBaconStorage.diskStorage: DiskStorage.sharedStorage
+        ]
     }
 
+    /**
+     Returns an adapter requested by its name
+     
+     - parameter name: the adapter name (use constants!)
+     
+     - returns: the adapter
+     */
+    private func adapter(name: String) -> Storage {
+        guard let adapter: Storage = self.adapters[name] else {
+            preconditionFailure("unknown adapter requested")
+        }
+        return adapter
+    }
 }
 
 extension MapleBaconStorage: Storage {
 
-    public func storeImage(image: UIImage, data: NSData?, forKey key: String) {
-        inMemoryStorage.storeImage(image, data: data, forKey: key)
-        diskStorage.storeImage(image, data: data, forKey: key)
+    public func store(image: UIImage, forKey key: String) {
+        
+        self.adapters.forEach { $1.store(image: image, forKey: key) }
+    }
+    
+    public func store(data: Data, forKey key: String) {
+        
+        self.adapters.forEach {$1.store(data: data, forKey: key) }
     }
     
     public func image(forKey key: String) -> UIImage? {
-        if let image = inMemoryStorage.image(forKey: key) {
-            return image
-        } else if let image = diskStorage.image(forKey: key) {
-            inMemoryStorage.storeImage(image, data: nil, forKey: key)
+        
+        // try reading from memory first
+        if let image: UIImage = self.adapter(name: MapleBaconStorage.inMemoryStorage).image(forKey: key) {
             return image
         }
-        return nil
+        
+        // read from all other if it fails
+        var img: UIImage? = nil
+        self.adapters.forEach {
```

Comment:
So maybe instead of all this dancing inside of `forEach` and the rich number of comments, remove in memory from the loop? `adapters.filter().forEach`